docs/external-helpers.md

```diff
@@ -0,0 +1,131 @@
+2. **Data Flow**: ZAP will transmit the necessary API object to these helpers at runtime during binary creation, allowing them to work with ZAP data without needing to directly interact with ZAP’s source code.
+3. **API Abstraction**: The API object passed to the helpers contains abstracted ZAP method definitions, meaning SDK developers do not need to worry about changes in query or method names within ZAP. If a query or method name changes, only the ZAP team needs to make updates to the API; the SDK code remains unaffected.
+
+### Example
+
+In the provided code snippet, the SDK receives an API object containing "attributes" which invokes a query. In the event of a query name change, the ZAP developer updates the API, relieving the SDK developer from making any adjustments:
+
+```javascript
+/**
+ * Returns an array of attributes for a given cluster.
+ * The cluster input is required to come from a script-api in this module.
+ *
+ * @param {*} context
+ * @param {*} endpoint
+ * @param {*} cluster
+ */
+function attributes(context, endpoint, cluster) {
+  return queryEndpoint.selectEndpointClusterAttributes(
+    context.db,
+    cluster.clusterId,
+    cluster.side,
+    endpoint.endpointTypeRef
+  )
+}
+```
+
+In this example, the attributes function makes use of the queryEndpoint.selectEndpointClusterAttributes method. If ZAP changes the method or its name, the SDK developers are not required to update their code. ZAP abstracts these changes through the API object passed to the helper, which remains stable.
+
+## How to Add Externalized Helpers
+
+The process for adding externalized helpers follows a new design pattern that enables SDK developers to extend functionality without modifying the ZAP repository.
+
+### Step 1: Define Helper Functions
+
+Your externalized helper functions should be asynchronous and accept the necessary data passed from ZAP, typically in the form of an api object. Here's an example of a helper that retrieves and counts available events:
+
+```javascript
+/**
+ * Helper function to get the total number of available events.
+ * @param {object} api - The API object.
+ * @returns {number} - The total number of events.
+ */
+async function get_total_events_helper(api) {
+  let events = await api.availableEvents(this)
+  let totalEvents = events.length
+  return totalEvents
+}
+```
+
+Each helper function can interact with the API object provided by ZAP to fetch and process data as needed.
+
+### Step 2: Specify Helpers in gen-template.json
+
+The gen-template.json file allows SDK developers to specify the path to the externalized helpers they wish to use. The file should include the helper file's relative path within your project structure.
+
+For example, your gen-template.json might include:
+
+```javascript
+{
+  "helpers": [
+    "path/to/external_helpers/get_total_events_helper.js",
```

Comment:
You should say explicitely that this is a relative path from the directory where the gen-template.json file itself is located.

Reply:
Okay will do